core/src/main/kotlin/supplier/CacheAwareRestSupplier.kt

```diff
@@ -0,0 +1,171 @@
+    }
+
+    override fun getGuildRoles(guildId: Snowflake): Flow<Role> {
+        return storeAndEmit(restSupplier.getGuildRoles(guildId)) { it.data }
+    }
+
+    override suspend fun getGuildBanOrNull(guildId: Snowflake, userId: Snowflake): Ban? {
+        return storeAndReturn(getGuildBanOrNull(guildId, userId)) { it.data }
+    }
+
+    override fun getGuildBans(guildId: Snowflake): Flow<Ban> {
+        return storeAndEmit(restSupplier.getGuildBans(guildId)) { it.data }
+    }
+
+    override fun getGuildMembers(guildId: Snowflake, limit: Int): Flow<Member> {
+        return storeAndEmit(restSupplier.getGuildMembers(guildId, limit)) { it.data }
+    }
+
+    override fun getGuildVoiceRegions(guildId: Snowflake): Flow<Region> {
+        return storeAndEmit(restSupplier.getGuildVoiceRegions(guildId)) { it.data }
+    }
+
+    override suspend fun getEmojiOrNull(guildId: Snowflake, emojiId: Snowflake): GuildEmoji? {
+        return storeAndReturn(restSupplier.getEmojiOrNull(guildId, emojiId)) { it.data }
+    }
+
+    override fun getEmojis(guildId: Snowflake): Flow<GuildEmoji> {
+        return storeAndEmit(restSupplier.getEmojis(guildId)) { it.data }
+
+    }
+
+    override fun getCurrentUserGuilds(limit: Int): Flow<Guild> {
+        return storeAndEmit(restSupplier.getCurrentUserGuilds(limit)) { it.data }
+
+    }
+
+    override fun getChannelWebhooks(channelId: Snowflake): Flow<Webhook> {
+        return storeAndEmit(restSupplier.getChannelWebhooks(channelId)) { it.data }
+    }
+
+    override fun getGuildWebhooks(guildId: Snowflake): Flow<Webhook> {
+        return storeAndEmit(restSupplier.getGuildWebhooks(guildId)) { it.data }
+    }
+
+    override suspend fun getWebhookOrNull(id: Snowflake): Webhook? {
+        return storeAndReturn(restSupplier.getWebhookOrNull(id)) { it.data }
+    }
+
+    override suspend fun getWebhookWithTokenOrNull(id: Snowflake, token: String): Webhook? {
+        return storeAndReturn(restSupplier.getWebhookWithTokenOrNull(id, token)) { it.data }
+    }
+
+    override suspend fun getTemplateOrNull(code: String): Template? {
+        return storeAndReturn(restSupplier.getTemplateOrNull(code)) { it.data }
+    }
+
+    override fun getTemplates(guildId: Snowflake): Flow<Template> {
+        return storeAndEmit(restSupplier.getTemplates(guildId)) { it.data }
+    }
+
+    override suspend fun getStageInstanceOrNull(channelId: Snowflake): StageInstance? {
+        return storeAndReturn(restSupplier.getStageInstanceOrNull(channelId)) { it.data }
+    }
+
+    private suspend inline fun <T, R> Flow<T>.mapAndStore(crossinline transform: (T) -> R) {
+        val data = this.map { transform(it) }
+        kord.cache.put(data)
+    }
+
+    private inline fun <T, R> storeAndEmit(source: Flow<T>, crossinline transform: (T) -> R): Flow<T> {
+        return flow {
+            source.mapAndStore { transform(it) }
+            emitAll(source)
+        }
+    }
```

Comment:
This iterates the flow twice, and iterates is before emitting any values. That's a major issue when the source is backed by a rest request (and possibly multiple given a paginator).

Reply:
We can emit all, then store. That way there will be no delay on the call site

Reply:
I've made the requested changes, and switched the method calls around. so that we emit before we store.